src/target/riscv/riscv.h

```diff
@@ -242,8 +242,11 @@ struct riscv_info {
 	int (*dmi_read)(struct target *target, uint32_t *value, uint32_t address);
 	int (*dmi_write)(struct target *target, uint32_t address, uint32_t value);
 
-	int (*dm_read)(struct target *target, uint32_t *value, uint32_t address);
-	int (*dm_write)(struct target *target, uint32_t address, uint32_t value);
+	/* Get the DMI address of target's DM's register.
+	 * The function should return the passed address
+	 * if the target is not assigned a DM yet.
```

Comment:
I am a bit uncertain about this part. I'd prefer to raise an error - to make sure the function cannot be used until the target is assigned a DM.

But this can be handled later (if and when).

Reply:
I agree with you. This change is a bit questionable.
The reason for the current solution is `get_dm()` function. Currently `get_dm()` function:
* Returns `NULL` if a DM was not allocated/selected prior and allocation failed (out of memory).
* Returns non-examined DM (the corresponding structure is just initialized, but is yet to be filled with correct values) if none of the targets, corresponding to this DM were examined.
* Returns a properly examined DM otherwise.

The second case is quite worrying. I would like to make this function a pure accessor and search for DM / allocate it explicitly in the `examine()`. I'm preparing the corresponding patch. This would make it easier to reason about current state of a DM and allow to replace this weird behavior of `get_dmi_address()` with an assertion.

Reply:
> I would like to make this function a pure accessor and search for DM / allocate it explicitly in the examine(). I'm preparing the corresponding patch.

That sounds very good.